### PR TITLE
add support for the putSecretValue restful API

### PR DESCRIPTION
**Pull request by Yidingluo1**
this restful API will be exposed to public as:
POST <ehsm_srv_address>/ehsm?Action=PutSecretValue

stores the secret value of a new version into a secret object.
by default, the newly stored secret value is marked with the current version stage, and the original versions
will be set as previously.

test: passed the unittest

Signed-off-by: yidingluo1 <luo.yd@neusoft.com>

### CODE REVIEW

File: ehsm_kms_service/secret_manager_apis.js [hunk 1]
```diff
@@ -265,9 +265,104 @@ const createSecret = async (res, appid, payload, DB) => {
...
+                'createTime',
+                'versionStage',
+            ],
+            limit: 1
+        }
+        const version_stage_res = await DB.partitionedFind('secret_version_data', query_version_stage)
+        if (version_stage_res.docs.length > 0) {
+            versionId = version_stage_res.docs[0].versionId + 1
+            version_stage_res.docs[0].versionStage = SM_SECRET_VERSION_STAGE_PREVIOUS
+            await DB.insert(version_stage_res.docs[0])
```
Comment (HawieLong) [inline ehsm_kms_service/secret_manager_apis.js:338]: use "insert" to update a DB? Can we use “update DB“ code here？

File: ehsm_kms_service/secret_manager_apis.js [hunk 2]
```diff
@@ -265,9 +265,104 @@ const createSecret = async (res, appid, payload, DB) => {
...
+        if (secret_metadata_res.docs.length == 0) {
+            encryptionKeyId = ""
+        } else {
+            encryptionKeyId = secret_metadata_res.docs[0].encryptionKeyId
+        }
+        retCMK = await get_SM_Masterkey(DB, appid, encryptionKeyId)
+        if (retCMK.isFind) {
+            sm_masterKey = retCMK.sm_masterKey
+        } else {
+            res.send(_result(400, retCMK.msg))
```
Comment (HawieLong) [inline ehsm_kms_service/secret_manager_apis.js:306]: If the user uses a new secretName to put, we still return CMK not find? We should confirm the secretName in DB first.

File: ehsm_kms_service/secret_manager_apis.js [hunk 3]
```diff
@@ -265,9 +265,104 @@ const createSecret = async (res, appid, payload, DB) => {
...
+        const query_encryptionKeyId = {
+            selector: {
+                appid,
+                secretName,
+            },
+            fields: ['encryptionKeyId'],
+            limit: 1,
+        }
+        const secret_metadata_res = await DB.partitionedFind('secret_metadata', query_encryptionKeyId)
+        if (secret_metadata_res.docs.length == 0) {
```
Comment (yang8621) [inline ehsm_kms_service/secret_manager_apis.js:297]: if secretName does not exist (CreateSecret has not ever been called before), the code will go into @298?

File: ehsm_kms_service/secret_manager_apis.js (hunk 2 above)
Comment (Yidingluo1) [reply ehsm_kms_service/secret_manager_apis.js:306]: no direct update function in CouchDB, can only be replaced by insert

File: ehsm_kms_service/secret_manager_apis.js (hunk 1 above)
Comment (Yidingluo1) [reply ehsm_kms_service/secret_manager_apis.js:338]: no direct update function in CouchDB, can only be replaced by insert

File: ehsm_kms_service/secret_manager_apis.js (hunk 2 above)
Comment (Yidingluo1) [reply ehsm_kms_service/secret_manager_apis.js:312]: problem fixed. now query the secretname first, and then find the encryptionKeyId exists

File: ehsm_kms_service/secret_manager_apis.js (hunk 3 above)
Comment (Yidingluo1) [reply ehsm_kms_service/secret_manager_apis.js:297]: problem fixed. now query the secretname first, and then find the encryptionKeyId exists

**Comment by syan10** [comment]
new commti msg
"
add support for the putSecretValue restful API

this restful API will be exposed to public as:
POST <ehsm_srv_address>/ehsm?Action=PutSecretValue

stores the secret value of a new version into a secret object.
by default, the newly stored secret value is marked with the current version stage, and the original versions
will be set as previously.

test: passed the unittest
"

**Comment by Yidingluo1** [comment]
> new commti msg " add support for the putSecretValue restful API
> 
> this restful API will be exposed to public as: POST <ehsm_srv_address>/ehsm?Action=PutSecretValue
> 
> stores the secret value of a new version into a secret object. by default, the newly stored secret value is marked with the current version stage, and the original versions will be set as previously.
> 
> test: passed the unittest "

commit changed

File: test/test_secret_manager_with_cli.py [hunk 4]
```diff
@@ -30,6 +30,8 @@ def test_secret_manager(base_url, headers):
 
     createSecret.createSecret(base_url, "secretData2", "secret002", None, "mysecret", "20d")
 
+    putSecretValue.putSecrectValue(base_url, "secret001", "secretData3")
```
Comment (syan10) [inline test/test_secret_manager_with_cli.py:33]: seems the param order not make sense, generally, if I'm a user for the secret, i will use it by the following cmd:

putSecretVaulle --secretName "xxxx" --secretData "xxx"

same with other APIs

Comment (Yidingluo1) [reply test/test_secret_manager_with_cli.py:33]: ok, I will submit a new PR later to modify the parameter order of the create secret test

File: ehsm_kms_service/secret_manager_apis.js [hunk 5]
```diff
@@ -265,9 +265,119 @@ const createSecret = async (res, appid, payload, DB) => {
...
+        }
+        retCMK = await get_SM_Masterkey(DB, appid, encryptionKeyId)
+        if (retCMK.isFind) {
+            sm_masterKey = retCMK.sm_masterKey
+        } else {
+            res.send(_result(400, retCMK.msg))
+            return
+        }
+        const apikey_encrypt_res = napi_result(cryptographic_apis.Encrypt, res, [sm_masterKey, secretData, ''])
+        const { ciphertext } = apikey_encrypt_res.result
```
Comment (yang8621) [inline ehsm_kms_service/secret_manager_apis.js:325]: will check any error returned from Encrypt?

File: ehsm_kms_service/secret_manager_apis.js (hunk 1 above)
Comment (yang8621) [reply ehsm_kms_service/secret_manager_apis.js:353]: if _id is matched, DB.insert will update the same record?  is my understanding correct?

File: ehsm_kms_service/secret_manager_apis.js [hunk 6]
```diff
@@ -265,9 +265,119 @@ const createSecret = async (res, appid, payload, DB) => {
...
+                '_rev',
+                'appid',
+                'secretName',
+                'versionId',
+                'deletedFlag',
+                'secretData',
+                'createTime',
+                'versionStage',
+            ],
+            limit: 1
```
Comment (yang8621) [inline ehsm_kms_service/secret_manager_apis.js:347]: How can make sure the only one result is the latest versionid?

File: ehsm_kms_service/secret_manager_apis.js (hunk 1 above)
Comment (yang8621) [reply ehsm_kms_service/secret_manager_apis.js:353]: Line 353 is to insert a new record OR to update the original one?

File: ehsm_kms_service/secret_manager_apis.js (hunk 6 above)
Comment (Yidingluo1) [reply ehsm_kms_service/secret_manager_apis.js:347]: always only one secret_version_data version stage is SM_SECRET_VERSION_STAGE_PREVIOUS

File: ehsm_kms_service/secret_manager_apis.js (hunk 5 above)
Comment (Yidingluo1) [reply ehsm_kms_service/secret_manager_apis.js:325]: will catch in line 370

**Pull request closed without merging**